whereby/sdk
# core: Implement joinRoom/leaveRoom actions

### Description

Implement `joinRoom()` and `leaveRoom()` actions on the `useRoomConnection` hook.

As part of this process this PR also refactors the join and leave logic in the SDK to support leaving and re-joining without needing to reload the SDK. Previously all join/leave logic was tied to the lifecycle of the useRoomConnection hook.

**Summary:**



**Related Issue:**



### Testing

1. Check out this branch
2. Run `yarn build && yarn dev`
3. In a browser tab, open the dev console and go to "Network" to view the websocket connections.
4. In the Storybook UI (@ http://localhost:6006/) open the story: "Custom UI" > "Room Connection Only"
5. After successfully joining a room you should see the signal server websocket connection as "pending" in the developer tools.
6. Click on the "leave" button should remove the user from the room.
7. The websocket connection in the dev tools should now be closed.
8. Click on the "Join room" button
9. The user should successfully re-join the room and a new signal server websocket should now be "pending"
10. In the Dev Tools > Network tab, change the network throttling option from "No throttling" to "Offline"
11. The UI should now display "Disconnected" while the network tab will show signal websocket connection trying to be re-established (unsuccessfully while the network is disconnected)
12. In the Dev Tools > Network tab, change the network throttling option back from "Offline" to "No throttling"
13. The user should successfully re-join the room automatically on successfully signal reconnect.

### Screenshots/GIFs (if applicable)



### Checklist

        under the terms of the MIT license.
